File: jas/document/document.py
"""Immutable document model.

A Document is an ordered list of Layers.

Elements within the document are identified by their path: a tuple of integer
indices tracing the route from the document's layer list to the element.
For example, path (0, 2) means layer 0, child 2.  Path (1,) means layer 1
itself.  This allows selections and updates without requiring element identity.
"""

from __future__ import annotations

import dataclasses
from collections.abc import Iterable
from dataclasses import dataclass, field
from typing import TypeVar

from geometry.element import Element, Group, Layer

_G = TypeVar("_G", bound=Group)

# A path identifies an element by its position in the document tree.
# Each integer is a child index at that level of the tree.
# (0,) -> layers[0]  (a Layer)
# (0, 2) -> layers[0].children[2]
# (0, 2, 1) -> layers[0].children[2].children[1]  (inside a group)
ElementPath = tuple[int, ...]


@dataclass(frozen=True)
class SortedCps:
    """Sorted, de-duplicated tuple of control-point indices.

    Invariant: the backing tuple is sorted ascending and contains no
    duplicates. All constructors and operations preserve it, so callers
    can rely on deterministic iteration order.
    """
    _indices: tuple[int, ...] = ()

    @staticmethod
    def from_iter(it: Iterable[int]) -> "SortedCps":
        return SortedCps(tuple(sorted(set(int(i) for i in it))))

    @staticmethod
    def single(i: int) -> "SortedCps":
        return SortedCps((int(i),))

    def __contains__(self, i: int) -> bool:
        # Linear scan is fine for the small sets we deal with; the
        # invariant guarantees the result is the same as binary search.
        return int(i) in self._indices

    def __iter__(self):
        return iter(self._indices)

    def __len__(self) -> int:
        return len(self._indices)

    def __bool__(self) -> bool:
        return bool(self._indices)

    def insert(self, i: int) -> "SortedCps":
        """Return a new SortedCps with `i` added; no-op if already present."""
        if int(i) in self._indices:
            return self
        return SortedCps.from_iter(self._indices + (int(i),))

    def symmetric_difference(self, other: "SortedCps") -> "SortedCps":
        return SortedCps.from_iter(set(self._indices) ^ set(other._indices))


@dataclass(frozen=True)
class _SelectionAll:
    """Marker for `SelectionKind.ALL` — every CP of the element selected."""
    pass


@dataclass(frozen=True)
class _SelectionPartial:
    """Subset of CPs selected (Partial Selection)."""
    cps: SortedCps


# Sum type: either `.all` or `.partial(SortedCps)`. We use a tagged
# dataclass pair instead of a true Enum so SortedCps payloads are easy
# to construct and pattern-match with `isinstance` checks.
SelectionKind = _SelectionAll | _SelectionPartial


def selection_all() -> SelectionKind:
    """Return the singleton `.all` selection kind."""
    return _SelectionAll()


def selection_partial(cps: Iterable[int]) -> SelectionKind:
    """Return a `.partial(SortedCps)` selection kind from any iterable."""
    return _SelectionPartial(SortedCps.from_iter(cps))


def selection_kind_contains(kind: SelectionKind, i: int) -> bool:
    """True if CP index `i` is selected. `.all` contains every index."""
    if isinstance(kind, _SelectionAll):
        return True
    return int(i) in kind.cps


def selection_kind_count(kind: SelectionKind, total: int) -> int:
    """Number of selected CPs. The caller supplies `total` so `.all` can answer."""
    if isinstance(kind, _SelectionAll):
        return total
    return len(kind.cps)


def selection_kind_is_all(kind: SelectionKind, total: int) -> bool:
    """True when every CP of an element with `total` CPs is selected."""
    if isinstance(kind, _SelectionAll):
        return True
    return len(kind.cps) == total


def selection_kind_to_sorted(kind: SelectionKind, total: int) -> SortedCps:
    """Materialize an explicit SortedCps for an element with `total` CPs."""
    if isinstance(kind, _SelectionAll):
        return SortedCps.from_iter(range(total))
    return kind.cps


# Per-element selection state: either the element is fully selected
# (`.all`) or only a subset of its control points are selected
# (`.partial`).  Hash/equality is by path only so that a frozenset
# behaves as a path-keyed collection.
@dataclass(frozen=True, eq=False)
class ElementSelection:
    path: ElementPath
    kind: SelectionKind = field(default_factory=selection_all)

    @staticmethod
    def all(path: ElementPath) -> "ElementSelection":
        return ElementSelection(path=path, kind=selection_all())

    @staticmethod
    def partial(path: ElementPath, cps: Iterable[int]) -> "ElementSelection":
        return ElementSelection(path=path, kind=selection_partial(cps))

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, ElementSelection):
            return NotImplemented
        return self.path == other.path

    def __hash__(self) -> int:
        return hash(self.path)

# A selection is an immutable set of ElementSelection entries (unique by path).
Selection = frozenset[ElementSelection]


@dataclass(frozen=True)
class Document:
    """A document consisting of an ordered list of layers and a selection."""
    layers: tuple[Layer, ...] = (Layer(),)
    selected_layer: int = 0
    selection: Selection = frozenset()

    def get_element_selection(self, path: ElementPath) -> ElementSelection | None:
        """Return the ElementSelection for the given path, or None."""
        for es in self.selection:
            if es.path == path:
                return es
        return None

    def selected_paths(self) -> frozenset[ElementPath]:
        """Return the set of all element paths in the selection."""
        return frozenset(es.path for es in self.selection)

    def bounds(self) -> tuple[float, float, float, float]:
        """Return the bounding box of all layers combined."""
        if not self.layers:
            return (0, 0, 0, 0)
        all_bounds = [layer.bounds() for layer in self.layers]
        min_x = min(b[0] for b in all_bounds)
        min_y = min(b[1] for b in all_bounds)
        max_x = max(b[0] + b[2] for b in all_bounds)
        max_y = max(b[1] + b[3] for b in all_bounds)
        return (min_x, min_y, max_x - min_x, max_y - min_y)

    def get_element(self, path: ElementPath) -> Element:
        """Return the element at the given path."""
        if not path:
            raise ValueError("Path must be non-empty")
        if path[0] < 0 or path[0] >= len(self.layers):
            raise ValueError(f"Layer index {path[0]} out of range (have {len(self.layers)} layers)")
        node: Element = self.layers[path[0]]
        for idx in path[1:]:
            if not isinstance(node, Group):
                raise ValueError(f"Expected Group at path, got {type(node).__name__}")
            if idx < 0 or idx >= len(node.children):
                raise ValueError(f"Child index {idx} out of range in path {path}")
            node = node.children[idx]
        return node

    def effective_visibility(self, path: ElementPath) -> "Visibility":
        """Return the effective visibility of the element at ``path``.

        Computed as the minimum of the visibilities of every element
        along the path from the root layer down to the target. A
        parent Group or Layer caps the visibility of everything it
        contains: if any ancestor is INVISIBLE, the result is
        INVISIBLE regardless of the target's own flag.
        """
        from geometry.element import Visibility
        if not path:
            return Visibility.PREVIEW
        if path[0] >= len(self.layers):
            return Visibility.PREVIEW
        node: Element = self.layers[path[0]]
        effective = node.visibility
        for idx in path[1:]:
            if not isinstance(node, Group):
                return effective
            if idx >= len(node.children):
                return effective
            node = node.children[idx]
            if node.visibility.value < effective.value:
                effective = node.visibility
        return effective

    def replace_element(self, path: ElementPath, new_elem: Element) -> "Document":
        """Return a new Document with the element at path replaced by new_elem."""
        if not path:
            raise ValueError("Path must be non-empty")
        new_layers = list(self.layers)
        if len(path) == 1:
            if not isinstance(new_elem, Layer):
                raise ValueError("Top-level element must be a Layer")
            new_layers[path[0]] = new_elem
        else:
            new_layers[path[0]] = _replace_in_group(self.layers[path[0]], path[1:], new_elem)
        return dataclasses.replace(self, layers=tuple(new_layers))


    def insert_element_after(self, path: ElementPath, new_elem: Element) -> "Document":
        """Return a new Document with new_elem inserted immediately after path."""
        if not path:
            raise ValueError("Path must be non-empty")
        if len(path) == 1:
            new_layers = list(self.layers)
            if not isinstance(new_elem, Layer):
                raise ValueError("Top-level element must be a Layer")
            new_layers.insert(path[0] + 1, new_elem)
            return dataclasses.replace(self, layers=tuple(new_layers))
        new_layers = list(self.layers)
        new_layers[path[0]] = _insert_after_in_group(
            self.layers[path[0]], path[1:], new_elem)
        return dataclasses.replace(self, layers=tuple(new_layers))


    def delete_element(self, path: ElementPath) -> "Document":
        """Return a new Document with the element at path removed."""
        if not path:
            raise ValueError("Path must be non-empty")
        new_layers = list(self.layers)
        if len(path) == 1:
            del new_layers[path[0]]
        else:
            new_layers[path[0]] = _remove_from_group(self.layers[path[0]], path[1:])
        return dataclasses.replace(self, layers=tuple(new_layers))

    def delete_selection(self) -> "Document":
        """Return a new Document with all selected elements removed and selection cleared."""
        doc = self
        # Sort paths in reverse so deletions don't shift earlier paths
        paths = sorted((es.path for es in self.selection), reverse=True)
        for path in paths:
            doc = doc.delete_element(path)
        return dataclasses.replace(doc, selection=frozenset())


def _expect_group(node: Element, context: str) -> Group:
    """Validate that node is a Group, raising ValueError if not."""
    if not isinstance(node, Group):
        raise ValueError(f"Expected Group {context}, got {type(node).__name__}")
    return node


def _remove_from_group(node: _G, rest: ElementPath) -> _G:
    """Remove the element at rest within a group."""
    new_children = list(node.children)
    if len(rest) == 1:
        del new_children[rest[0]]
    else:
        child = _expect_group(node.children[rest[0]], "in path")
        new_children[rest[0]] = _remove_from_group(child, rest[1:])
    return dataclasses.replace(node, children=tuple(new_children))


def _insert_after_in_group(node: _G, rest: ElementPath, new_elem: Element) -> _G:
    """Insert new_elem after the position indicated by rest within a group."""
    new_children = list(node.children)
    if len(rest) == 1:
        new_children.insert(rest[0] + 1, new_elem)
    else:
        child = _expect_group(node.children[rest[0]], "in path")
        new_children[rest[0]] = _insert_after_in_group(child, rest[1:], new_elem)
    return dataclasses.replace(node, children=tuple(new_children))


def _replace_in_group(node: _G, rest: ElementPath, new_elem: Element) -> _G:
    """Recursively replace the element at rest within a group, returning the same Group subtype."""
    new_children = list(node.children)
    if len(rest) == 1:
        new_children[rest[0]] = new_elem
    else:
        child = _expect_group(node.children[rest[0]], "in path")
        new_children[rest[0]] = _replace_in_group(child, rest[1:], new_elem)
    return dataclasses.replace(node, children=tuple(new_children))

File: jas_ocaml/lib/document/document.ml
(** Immutable document model.

    A document is an ordered array of layers.

    Elements within the document are identified by their path: a list of integer
    indices tracing the route from the document's layer array to the element.
    For example, path [0; 2] means layer 0, child 2.  Path [1] means layer 1
    itself.  This allows selections and updates without requiring element identity. *)

open Element

(** A path identifies an element by its position in the document tree. *)
type element_path = int list

(** A selection is a set of element paths. *)
module PathSet = Set.Make(struct
  type t = element_path
  let compare = compare
end)

(** Sorted, de-duplicated list of control-point indices.

    The wrapper enforces the invariant: the backing list is sorted
    ascending and contains no duplicates. All constructors and
    operations preserve it. *)
module SortedCps = struct
  type t = int list  (* sorted, unique *)

  let empty : t = []

  let from_list (xs : int list) : t =
    List.sort_uniq compare xs

  let single i : t = [i]

  let mem i (s : t) = List.mem i s

  let to_list (s : t) : int list = s

  let length (s : t) = List.length s

  let is_empty (s : t) = s = []

  let insert i (s : t) : t =
    if List.mem i s then s else from_list (i :: s)

  let symmetric_difference (a : t) (b : t) : t =
    let in_a_only = List.filter (fun x -> not (List.mem x b)) a in
    let in_b_only = List.filter (fun x -> not (List.mem x a)) b in
    from_list (in_a_only @ in_b_only)
end

(** Per-element selection kind: either the element is fully selected
    (`SelKindAll`) or only a subset of its CPs are selected
    (`SelKindPartial of SortedCps.t`). *)
type selection_kind =
  | SelKindAll
  | SelKindPartial of SortedCps.t

let selection_kind_contains kind i =
  match kind with
  | SelKindAll -> true
  | SelKindPartial s -> SortedCps.mem i s

let selection_kind_count kind ~total =
  match kind with
  | SelKindAll -> total
  | SelKindPartial s -> SortedCps.length s

let selection_kind_is_all kind ~total =
  match kind with
  | SelKindAll -> true
  | SelKindPartial s -> SortedCps.length s = total

let selection_kind_to_sorted kind ~total =
  match kind with
  | SelKindAll ->
    let rec range i n = if i >= n then [] else i :: range (i + 1) n in
    range 0 total
  | SelKindPartial s -> SortedCps.to_list s

(** Per-element selection state. *)
type element_selection = {
  es_path : element_path;
  es_kind : selection_kind;
}

(** A selection is a map from element path to its selection state. *)
module PathMap = Map.Make(struct
  type t = element_path
  let compare = compare
end)

type selection = element_selection PathMap.t

(** A document consisting of an ordered array of layers. *)
type document = {
  layers : element array;
  selected_layer : int;
  selection : selection;
}

let make_document ?(selected_layer = 0) ?(selection = PathMap.empty) layers =
  { layers; selected_layer; selection }

(** Build a fully-selected entry for [path]. *)
let element_selection_all path =
  { es_path = path; es_kind = SelKindAll }

(** Build a partial entry for [path] from a list of CP indices. *)
let element_selection_partial path cps =
  { es_path = path; es_kind = SelKindPartial (SortedCps.from_list cps) }

(** Legacy constructor kept for compatibility with existing call sites. *)
let make_element_selection ?(control_points = []) path =
  if control_points = [] then element_selection_all path
  else element_selection_partial path control_points

(** Return the set of all element paths in the selection. *)
let selected_paths sel =
  PathMap.fold (fun path _ acc -> PathSet.add path acc) sel PathSet.empty

(** Return the element_selection for the given path, or None. *)
let get_element_selection sel path =
  PathMap.find_opt path sel

let default_document () = make_document
  ~selection:PathMap.empty [| Element.make_layer [||] |]

let bounds doc =
  if Array.length doc.layers = 0 then (0.0, 0.0, 0.0, 0.0)
  else
    let all_bounds = Array.map Element.bounds doc.layers in
    let min_x = Array.fold_left (fun acc (x, _, _, _) -> min acc x) infinity all_bounds in
    let min_y = Array.fold_left (fun acc (_, y, _, _) -> min acc y) infinity all_bounds in
    let max_x = Array.fold_left (fun acc (x, _, w, _) -> max acc (x +. w)) neg_infinity all_bounds in
    let max_y = Array.fold_left (fun acc (_, y, _, h) -> max acc (y +. h)) neg_infinity all_bounds in
    (min_x, min_y, max_x -. min_x, max_y -. min_y)

(** Return the children of an element, if it is a group or layer. *)
let children_of = function
  | Group { children; _ } | Layer { children; _ } -> children
  | _ -> failwith "element has no children"

(** Return the element at the given path in the document. *)
let get_element doc path =
  match path with
  | [] -> failwith "path must be non-empty"
  | [i] -> doc.layers.(i)
  | i :: rest ->
    let rec walk node = function
      | [] -> node
      | j :: rest -> walk (children_of node).(j) rest
    in
    walk doc.layers.(i) rest

let effective_visibility doc path =
  match path with
  | [] -> Element.Preview
  | i :: rest ->
    if i >= Array.length doc.layers then Element.Preview
    else
      let node = ref doc.layers.(i) in
      let effective = ref (Element.get_visibility !node) in
      let rec walk = function
        | [] -> ()
        | j :: rest ->
          let children = children_of !node in
          if j >= Array.length children then ()
          else begin
            node := children.(j);
            let v = Element.get_visibility !node in
            if compare v !effective < 0 then effective := v;
            walk rest
          end
      in
      walk rest;
      !effective

(** Return a copy of the array with element at index n replaced. *)
let array_replace_nth arr n x =
  let a = Array.copy arr in
  a.(n) <- x; a

(** Return the node with new_children substituted. *)
let with_children node new_children =
  match node with
  | Group { opacity; transform; locked; visibility; _ } ->
    Group { children = new_children; opacity; transform; locked; visibility }
  | Layer { name; opacity; transform; locked; visibility; _ } ->
    Layer { name; children = new_children; opacity; transform; locked; visibility }
  | _ -> failwith "element has no children"

(** Recursively replace the element at [rest] inside a group/layer node. *)
let rec replace_in_group node rest new_elem =
  match rest with
  | [] -> new_elem
  | [i] ->
    with_children node (array_replace_nth (children_of node) i new_elem)
  | i :: rest ->
    let cs = children_of node in
    let new_child = replace_in_group cs.(i) rest new_elem in
    with_children node (array_replace_nth cs i new_child)

(** Return a new array with x inserted after position n. *)
let array_insert_after arr n x =
  let len = Array.length arr in
  Array.init (len + 1) (fun i ->
    if i <= n then arr.(i)
    else if i = n + 1 then x
    else arr.(i - 1))

(** Recursively insert new_elem after the position indicated by rest. *)
let rec insert_after_in_group node rest new_elem =
  match rest with
  | [] -> failwith "rest must be non-empty"
  | [i] ->
    with_children node (array_insert_after (children_of node) i new_elem)
  | i :: rest ->
    let cs = children_of node in
    let new_child = insert_after_in_group cs.(i) rest new_elem in
    with_children node (array_replace_nth cs i new_child)

(** Return a new document with new_elem inserted immediately after path. *)
let insert_element_after doc path new_elem =
  match path with
  | [] -> failwith "path must be non-empty"
  | [i] ->
    { doc with layers = array_insert_after doc.layers i new_elem }
  | i :: rest ->
    let layer = doc.layers.(i) in
    let new_layer = insert_after_in_group layer rest new_elem in
    { doc with layers = array_replace_nth doc.layers i new_layer }

(** Return a new document with the element at [path] replaced by [new_elem]. *)
let replace_element doc path new_elem =
  match path with
  | [] -> failwith "path must be non-empty"
  | [i] ->
    { doc with layers = array_replace_nth doc.layers i new_elem }
  | i :: rest ->
    let layer = doc.layers.(i) in
    let new_layer = replace_in_group layer rest new_elem in
    { doc with layers = array_replace_nth doc.layers i new_layer }

(** Return a new array with element at index n removed. *)
let array_remove_nth arr n =
  let len = Array.length arr in
  Array.init (len - 1) (fun i -> if i < n then arr.(i) else arr.(i + 1))

(** Recursively remove the element at [rest] inside a group/layer node. *)
let rec remove_from_group node rest =
  match rest with
  | [] -> failwith "rest must be non-empty"
  | [i] -> with_children node (array_remove_nth (children_of node) i)
  | i :: rest ->
    let cs = children_of node in
    let new_child = remove_from_group cs.(i) rest in
    with_children node (array_replace_nth cs i new_child)

(** Return a new document with the element at [path] removed. *)
let delete_element doc path =
  match path with
  | [] -> failwith "path must be non-empty"
  | [i] -> { doc with layers = array_remove_nth doc.layers i }
  | i :: rest ->
    let layer = doc.layers.(i) in
    let new_layer = remove_from_group layer rest in
    { doc with layers = array_replace_nth doc.layers i new_layer }

(** Return a new document with all selected elements removed and selection cleared. *)
let delete_selection doc =
  let paths = PathMap.fold (fun _ es acc -> es.es_path :: acc) doc.selection [] in
  let sorted_paths = List.sort (fun a b -> compare b a) paths in
  let doc' = List.fold_left (fun d p -> delete_element d p) doc sorted_paths in
  { doc' with selection = PathMap.empty }
